### PR TITLE
added solution

### PR DESCRIPTION
**Pull request by anatolii-zarytskyi**

### CODE REVIEW

File: src/main/java/mate/academy/dao/impl/UserDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            return session.createQuery("FROM User u "
+                    + " WHERE u.email = :email", User.class)
+                    .setParameter("email", email)
+                    .uniqueResultOptional();
+        } catch (Exception e) {
+            throw new DataProcessingException("Can't find user by email: " + email, e);
+        }
+    }
+}
```
Comment (ritakisluhina) [inline src/main/java/mate/academy/dao/impl/UserDaoImpl.java:49]: I think we also need get method in UserDaoImpl

File: src/main/java/mate/academy/model/User.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    public void setSalt(byte[] salt) {
+        this.salt = salt;
+    }
+
+    @Override
+    public String toString() {
+        return "User{"
+                + "id=" + id
+                + ", email='" + email + '\''
+                + ", password='" + password + '\''
```
Comment (ritakisluhina) [inline src/main/java/mate/academy/model/User.java:58]: is it good to show the password?

File: src/main/java/mate/academy/security/AuthenticationServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+
+    @Override
+    public User register(String email, String password) throws AuthenticationException {
+        if (userService.findByEmail(email).isPresent()) {
+            throw new AuthenticationException("User with this email already exists: " + email);
+        }
+        User user = new User();
+        user.setEmail(email);
+        user.setPassword(password);
+        return userService.add(user);
```
Comment (ritakisluhina) [inline src/main/java/mate/academy/security/AuthenticationServiceImpl.java:35]: maybe it is better to use constructor? but maybe not ^_^

File: src/main/java/mate/academy/util/HashUtil.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        try {
+            MessageDigest messageDigest = MessageDigest.getInstance(CRYPTO_ALGORITHM);
+            messageDigest.update(salt);
+            byte[] digest = messageDigest.digest(password.getBytes());
+            for (byte b : digest) {
+                hashedPassword.append(String.format("%02x", b));
+            }
+            return hashedPassword.toString();
+        } catch (NoSuchAlgorithmException e) {
+            throw new IllegalStateException("Could not create hash using SHA-512 algorithm", e);
```
Comment (ritakisluhina) [inline src/main/java/mate/academy/util/HashUtil.java:31]: we have the constant CRYPTO_ALGORITHM, use it here

**Review by ritakisluhina — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/model/User.java (hunk 2 above)
Comment (Anatolii-Zarytskyi) [reply src/main/java/mate/academy/model/User.java:58]: agree

File: src/main/java/mate/academy/util/HashUtil.java (hunk 4 above)
Comment (Anatolii-Zarytskyi) [reply src/main/java/mate/academy/util/HashUtil.java:31]: done

File: src/main/java/mate/academy/security/AuthenticationServiceImpl.java (hunk 3 above)
Comment (Anatolii-Zarytskyi) [reply src/main/java/mate/academy/security/AuthenticationServiceImpl.java:35]: maybe not)

File: src/main/java/mate/academy/dao/impl/UserDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't insert user: " + user + " to DB", e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
+        return user;
```
[the remark this answers is not in the document]
Comment (Anatolii-Zarytskyi) [reply src/main/java/mate/academy/dao/impl/UserDaoImpl.java:35]: done

File: src/main/java/mate/academy/dao/impl/UserDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+@Dao
+public class UserDaoImpl implements UserDao {
+    @Override
+    public User add(User user) {
+        SessionFactory sessionFactory = HibernateUtil.getSessionFactory();
+        Session session = null;
+        Transaction transaction = null;
+        try {
+            session = sessionFactory.openSession();
```
[the remark this answers is not in the document]
Comment (Anatolii-Zarytskyi) [reply src/main/java/mate/academy/dao/impl/UserDaoImpl.java:21]: done

**Review by ritakisluhina — APPROVED** [review]

File: src/main/java/mate/academy/security/AuthenticationServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    @Inject
+    private UserService userService;
+
+    @Override
+    public User login(String email, String password) throws AuthenticationException {
+        Optional<User> userOptional = userService.findByEmail(email);
+        if (userOptional.isPresent() && userOptional.get().getPassword()
+                .equals(HashUtil.hashPassword(password, userOptional.get().getSalt()))) {
+            return userOptional.get();
+
```
Comment (Resci) [inline src/main/java/mate/academy/security/AuthenticationServiceImpl.java:22]:
```suggestion
```

File: src/main/resources/hibernate.cfg.xml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+        <property name="connection.password">101351013</property>
+        <property name="show_sql">true</property>
+        <property name="hbm2ddl.auto">create-drop</property>
+
+        <mapping class="mate.academy.model.Movie"/>
+        <mapping class="mate.academy.model.CinemaHall"/>
+        <mapping class="mate.academy.model.MovieSession"/>
+        <mapping class="mate.academy.model.User"/>
+    </session-factory>
+</hibernate-configuration>
```
Comment (Resci) [inline src/main/resources/hibernate.cfg.xml:21]: Add empty line

**Review by Resci — APPROVED** [review]

**Pull request closed without merging**